### PR TITLE
model: LFM2-VL fixes

### PR DESCRIPTION
**Pull request by tdakhran**
Debugging of #17290 revealed multiple issues with LFM2-VL.

This PR fixes the following issues and makes the output of `llama.cpp` equivalent to PyTorch
* surround image embeddings with <|image_start|> and <|image_end|> tokens
* use `round_by_factor` to calculate target `width` and `height` in "smart resize".
* stretch the image to the width and height calculated by smart resize instead of padding
* place image embeddings before user prompt
* resize positional embedding with antialiasing enabled

The central issue was the resizing of positional embeddings. Siglip2 implementation in PyTorch uses `F.interpolate(..., mode="bilinear", align_corners=False, antialias=True)`. Antialiasing only contributes during downscaling. When the image width or height is less than 256, the scaling of positional embeddings in `llama.cpp` produced numerically different results from PyTorch.

A new flag, ' GGML_SCALE_FLAG_ANTIALIAS', has been added for the upscale function, with implementations for CPU and CUDA.
Now outputs match:
siglip_1024
PyTorch (fp32)
```
For the vision tower, LF2-M2-VL uses Sigilip2 NaFlex encoders to convert input images into token sequences. Two variants are implemented:
```
this PR (`bin/llama-mtmd-cli -m $CKPT/LFM2-VL-1.6B-F32.gguf  --mmproj $CKPT/mmproj-LFM2-VL-1.6B-F32.gguf -n 64 -t 4 --image /data/playground/issue_17290/siglip_1024.png -p "OCR." --temp 0.0 --top-k 1`)
```
For the vision tower, LF2-M2-VL uses Sigilip2 NaFlex encoders to convert input images into token sequences. Two variants are implemented:
```

### CODE REVIEW

File: ggml/src/ggml-cpu/ops.cpp [hunk 1]
```diff
@@ -7420,6 +7420,65 @@ static void ggml_compute_forward_upscale_f32(
                 }
             }
         }
+    } else if (mode == GGML_SCALE_MODE_BILINEAR && (mode_flags & GGML_SCALE_FLAG_ANTIALIAS)) {
```
Comment (tdakhran) [inline ggml/src/ggml-cpu/ops.cpp:7423]: I was thinking whether to introduce a new scaling mode `GGML_SCALE_MODE_BILINEAR_ANTIALIAS` or a flag. I would like to hear your feedback on this.
Another question is where to place correctness tests?

File: ggml/src/ggml-cpu/ops.cpp [hunk 2]
```diff
@@ -7420,6 +7420,65 @@ static void ggml_compute_forward_upscale_f32(
...
+        // Similar to F.interpolate(..., mode="bilinear", align_corners=False, antialias=True)
+        // https://github.com/pytorch/pytorch/blob/8871ff29b743948d1225389d5b7068f37b22750b/aten/src/ATen/native/cpu/UpSampleKernel.cpp
+        auto triangle_filter = [](float x) -> float {
+            return std::max(1.0f - fabsf(x), 0.0f);
+        };
+
+        // support and invscale, maximum 1 pixel for bilinear
```
Comment (tdakhran) [inline ggml/src/ggml-cpu/ops.cpp]: shall be minimum in the comment

File: ggml/src/ggml-cuda/upscale.cu [hunk 3]
```diff
@@ -81,6 +81,76 @@ static __global__ void upscale_f32_bilinear(const float * x, float * dst,
...
+    const int i12_dst = (index / (ne10_dst * ne11_dst)) % ne12_dst;
+    const int i13_dst = index / (ne10_dst * ne11_dst * ne12_dst);
+
+    const int i02_src = (int)(i12_dst / sf2);
+    const int i03_src = (int)(i13_dst / sf3);
+
+    const float y = ((float)i11_dst + pixel_offset) / sf1;
+    const float x = ((float)i10_dst + pixel_offset) / sf0;
+
+    // support and invscale, maximum 1 pixel for bilinear
```
Comment (tdakhran) [inline ggml/src/ggml-cuda/upscale.cu]: shall be minimum in the comment

File: ggml/src/ggml-cuda/upscale.cu [hunk 4]
```diff
@@ -207,9 +281,10 @@ void ggml_cuda_op_upscale(ggml_backend_cuda_context & ctx, ggml_tensor * dst) {
     if (mode == GGML_SCALE_MODE_NEAREST) {
         upscale_f32_cuda(src0_d, dst_d, src0->nb[0], src0->nb[1], src0->nb[2], src0->nb[3], dst->ne[0], dst->ne[1], dst->ne[2], dst->ne[3], sf0, sf1, sf2, sf3, stream);
     } else if (mode == GGML_SCALE_MODE_BILINEAR) {
+        bool antialias = (mode_flags & GGML_SCALE_FLAG_ANTIALIAS);
```
Comment (tdakhran) [inline ggml/src/ggml-cuda/upscale.cu]: make const

File: tools/mtmd/mtmd.h [hunk 5]
```diff
@@ -299,6 +299,9 @@ struct input_chunks {
     }
 };
 
+// insert mtmd_default_marker() into given string, position depends on the projector
+std::string mtmd_add_default_marker(mtmd_context *ctx, const std::string &str);
```
Comment (ngxson) [inline tools/mtmd/mtmd.h]: I'd prefer to remove API because:
- It's not compatible with pure-C ABI
- The ordering is actually controlled by users via API. This function only change `llama-mtmd-cli`, but make no changes to `llama-server`

File: tools/mtmd/clip.cpp [hunk 6]
```diff
@@ -2787,7 +2787,7 @@ struct clip_model_loader {
                     {
                         get_u32(KEY_PROJ_SCALE_FACTOR, hparams.n_merge, false);
                         // ref: https://huggingface.co/LiquidAI/LFM2-VL-3B/blob/main/preprocessor_config.json
-                        hparams.set_limit_image_tokens(64, 256);
+                        hparams.set_limit_image_tokens(64, 1024);
```
Comment (tdakhran) [inline tools/mtmd/clip.cpp:2791]: Add a comment that in the config, the number of tokens is after downsampling, while here it is before.

File: tools/mtmd/mtmd.h (hunk 5 above)
Comment (ngxson) [reply tools/mtmd/mtmd.h]: A function like `mtmd_get_image_placement` can be a better solution, it returns one of these 3 values which should cover all use cases possible:

```cpp
IMAGE_PLACEMENT_NONE, // place images freely inside the message
IMAGE_PLACEMENT_BEGIN, // place images in the beginning of the image
IMAGE_PLACEMENT_END, // place in the end
```

But IMO this can better be a dedicated refactoring

Comment (tdakhran) [reply tools/mtmd/mtmd.h]: I'll remove it from this PR. For now, passing a placeholder directly in `-p "<__media__>OCR." ` achieves the same. But I wanted the placement to be correct by default.

Comment (tdakhran) [reply tools/mtmd/mtmd.h]: My understanding is that for the server, the order is similar to the order of data in the request, for the CLI, it was using `content += mtmd_default_marker();`.

Comment (ngxson) [reply tools/mtmd/mtmd.h]: For CLI, it was done this way because you can actually add multiple images in CLI mode, something like this:

```
> This is the first step:
> /image step1.png
> Then the next step:
> /image step2.png
> What do you see?
```

In this case, we expect the image and text prompts to follow exactly the same order in the input.

File: ggml/src/ggml-cpu/ops.cpp (hunk 1 above)
Comment (ngxson) [reply ggml/src/ggml-cpu/ops.cpp:7423]: I think it's fine to introduce this `GGML_SCALE_FLAG_ANTIALIAS` as a flag. Btw, I think we would need to explicitly `GGML_ASSERT` that the antialias is only supported by bilinear (maybe add a TODO to implement it in other modes)

**Review by ngxson — COMMENTED** [review]
I think we probably need to update `ggml_backend_*_supports_op` across backend, to avoid some backend fallback to the non-antialias kernel, which will result in wrong results.

For backend that does not support this mode, it will be fallback to CPU

**Comment by SmartestWashingMachine** [comment]
On my end (CPU) the outputs of fp32 and bf16 450M looked good, tested on a variety of small images (< 16/32px one side). 

Also checked a few personal tuned 1.6B Q3s (which should be more sensitive) and the outputs were great - it didn't go into a repetitive "breaking" state like before!

It couldn't have been easy to figure this issue out... Thank you guys for looking into this!

File: ggml/src/ggml-cpu/ops.cpp (hunk 1 above)
Comment (tdakhran) [reply ggml/src/ggml-cpu/ops.cpp:7423]: added assert in https://github.com/ggml-org/llama.cpp/pull/17577/commits/b81928f9795da0a7cb8af1592cb018a2e59f1ba6

File: tools/mtmd/mtmd.h (hunk 5 above)
Comment (tdakhran) [reply tools/mtmd/mtmd.h]: Rolled back changes in https://github.com/ggml-org/llama.cpp/pull/17577/commits/31be1a9fd11fadfa1d5df9bc0709d5f38ba40137, will address them in separate PR

**Comment by tdakhran** [comment]
@SmartestWashingMachine, great to see it work for you!

@ngxson, I [rolled back changes](https://github.com/ggml-org/llama.cpp/pull/17577/commits/31be1a9fd11fadfa1d5df9bc0709d5f38ba40137) related to default marker placement, for now `-p "<__media__>OCR."` has to be explicitly specified for LFM2-VL for `llama-mtmd-cli`, otherwise incorrect output will be produced.

Added a fallback to CPU for backends as well.

File: tools/mtmd/mtmd.h (hunk 5 above)
Comment (tdakhran) [reply tools/mtmd/mtmd.h]: Implemented in https://github.com/ggml-org/llama.cpp/pull/17616

File: tools/mtmd/clip.cpp [hunk 7]
```diff
@@ -4357,7 +4359,8 @@ bool clip_image_preprocess(struct clip_ctx * ctx, const clip_image_u8 * img, str
                 const std::array<uint8_t, 3> pad_color = {122, 116, 104};
 
                 clip_image_u8 resized_img;
-                img_tool::resize(*img, resized_img, target_size, img_tool::RESIZE_ALGO_BILINEAR, true, pad_color);
+                const bool pad = (ctx->proj_type() != PROJECTOR_TYPE_LFM2);
```
Comment (ngxson) [inline tools/mtmd/clip.cpp:4362]: self-note: probably kimi-vl also need padding, to be verified later

**Review by ngxson — COMMENTED** [review]
The mtmd changes look good to me. I'll also need a review on backend code (cpu/cuda) before merging this. CC @CISC @JohannesGaessler if you have a bit of time.

**Review by JohannesGaessler — APPROVED** [review]
The CUDA changes look good to me.

**Review by CISC — APPROVED** [review]
